llama_index/callbacks/open_inference_callback.py

```diff
@@ -0,0 +1,244 @@
+"""
+Callback handler for storing generation data in OpenInference format.
+OpenInference is an open standard for capturing and storing AI model inferences.
+It enables production LLMapp servers to seamlessly integrate with LLM
+observability solutions such as Arize and Phoenix.
+
+For more information on the specification, see
+https://github.com/Arize-ai/open-inference-spec
+"""
+
+import importlib
+import uuid
+from dataclasses import dataclass, field, fields
+from datetime import datetime
+from types import ModuleType
+from typing import TYPE_CHECKING, Any, Callable, Dict, Iterable, List, Optional, TypeVar
+
+from llama_index.callbacks.base import BaseCallbackHandler
+from llama_index.callbacks.schema import CBEventType, EventPayload
+
+if TYPE_CHECKING:
+    from pandas import DataFrame
+
+
+OPENINFERENCE_COLUMN_NAME = "openinference_column_name"
+Embedding = List[float]
+
+
+def _generate_random_id() -> str:
+    """Generates a random ID.
+
+    Returns:
+        str: A random ID.
+    """
+    return str(uuid.uuid4())
+
+
+@dataclass
+class QueryData:
+    """
+    Query data with column names following the OpenInference specification.
+    """
+
+    id: str = field(
+        default_factory=_generate_random_id,
+        metadata={OPENINFERENCE_COLUMN_NAME: ":id.str:"},
+    )
+    timestamp: Optional[str] = field(
+        default=None, metadata={OPENINFERENCE_COLUMN_NAME: ":timestamp.iso_8601:"}
+    )
+    query_text: Optional[str] = field(
+        default=None,
+        metadata={OPENINFERENCE_COLUMN_NAME: ":feature.text:prompt"},
+    )
+    query_embedding: Optional[Embedding] = field(
+        default=None,
+        metadata={OPENINFERENCE_COLUMN_NAME: ":feature.[float].embedding:prompt"},
+    )
+    response_text: Optional[str] = field(
+        default=None, metadata={OPENINFERENCE_COLUMN_NAME: ":prediction.text:response"}
+    )
+    document_ids: List[str] = field(
+        default_factory=list,
+        metadata={
+            OPENINFERENCE_COLUMN_NAME: ":feature.[str].retrieved_document_ids:prompt"
+        },
+    )
+    scores: List[float] = field(
+        default_factory=list,
+        metadata={
+            OPENINFERENCE_COLUMN_NAME: (
+                ":feature.[float].retrieved_document_scores:prompt"
```

Comment:
Similar to above, these are actually node ids. Up to you though, I'm unsure if this naming is dependent on upstream consumers or not :)

Reply:
The "document" terminology here is part of the spec for OpenInference.